modules/apps/frontend-data-set/frontend-data-set-admin-web/src/main/java/com/liferay/frontend/data/set/admin/web/internal/model/DataSetModelManagerImpl.java

```diff
@@ -0,0 +1,1013 @@
+import com.liferay.object.model.ObjectField;
+import com.liferay.object.model.ObjectFolder;
+import com.liferay.object.service.ObjectDefinitionLocalService;
+import com.liferay.object.service.ObjectFieldLocalService;
+import com.liferay.object.service.ObjectFolderLocalService;
+import com.liferay.object.service.ObjectRelationshipLocalService;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.dao.orm.QueryUtil;
+import com.liferay.portal.kernel.exception.PortalException;
+import com.liferay.portal.kernel.feature.flag.FeatureFlagManagerUtil;
+import com.liferay.portal.kernel.language.Language;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.model.Company;
+import com.liferay.portal.kernel.model.User;
+import com.liferay.portal.kernel.module.util.BundleUtil;
+import com.liferay.portal.kernel.service.CompanyLocalService;
+import com.liferay.portal.kernel.service.UserLocalService;
+import com.liferay.portal.kernel.util.Portal;
+import com.liferay.portal.kernel.workflow.WorkflowConstants;
+import com.liferay.portal.vulcan.util.LocalizedMapUtil;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Daniel Sanz
+ */
+@Component(service = DataSetModelManager.class)
+public class DataSetModelManagerImpl implements DataSetModelManager {
+
+	public void checkCompany(long companyId) {
+		try {
+			if (_checkedCompanies == null) {
+				_checkedCompanies = new HashSet<>();
+			}
+
+			if (_checkedCompanies.contains(companyId)) {
+				return;
+			}
+
+			ObjectDefinition fdsViewObjectDefinition =
+				_objectDefinitionLocalService.fetchObjectDefinition(
+					companyId, "FDSView");
+
+			if (fdsViewObjectDefinition != null) {
+				_checkedCompanies.add(companyId);
+
+				return;
+			}
+
+			generate(_companyLocalService.getCompany(companyId));
+		}
+		catch (PortalException portalException) {
+			_log.error(portalException);
+		}
+	}
+
+	public void generate(Company company) {
+		try {
+			BatchEngineUnitThreadLocal.setFileName(_bundle.toString());
+
+			if (_log.isInfoEnabled()) {
+				_log.info("Generating data set model for " + company.getName());
+			}
+
+			List<User> users = _userLocalService.getCompanyUsers(
+				company.getCompanyId(), QueryUtil.ALL_POS, QueryUtil.ALL_POS);
+
+			User adminUser = null;
+
+			for (User user : users) {
+				if (_portal.isCompanyAdmin(user)) {
+					adminUser = user;
+
+					break;
+				}
+			}
+
+			if (adminUser == null) {
+				throw new PortalException(
+					"Unable to find administrator user for company " +
+						company.getCompanyId());
+			}
```

Comment:
Are we sure that there will be a company admin at the time of company creation?

Reply:
I think so because of the transaction commit callback. Still we need to find an admin user and fail if there is not any.